storm-server/src/main/java/org/apache/storm/metric/StormMetricsRegistry.java

```diff
@@ -12,28 +12,30 @@
 
 package org.apache.storm.metric;
 
+import com.codahale.metrics.ExponentiallyDecayingReservoir;
 import com.codahale.metrics.Gauge;
 import com.codahale.metrics.Histogram;
 import com.codahale.metrics.Meter;
 import com.codahale.metrics.Metric;
 import com.codahale.metrics.MetricRegistry;
 import com.codahale.metrics.MetricSet;
 import com.codahale.metrics.Reservoir;
+
 import java.util.Map;
 import java.util.concurrent.Callable;
+
+import com.codahale.metrics.Timer;
+import org.apache.commons.lang.StringUtils;
 import org.apache.storm.daemon.metrics.MetricsUtils;
 import org.apache.storm.daemon.metrics.reporters.PreparableReporter;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-@SuppressWarnings("unchecked")
-public class StormMetricsRegistry {
-    private static final MetricRegistry DEFAULT_REGISTRY = new MetricRegistry();
+public class StormMetricsRegistry extends MetricRegistry {
```

Comment:
I switched to inheritance because I need to override parent's `register()` method and swallows the exceptions, instead of provide a static `register()` method which delegate the registration. This is because dropwizard handles registering MetricSet differently under the hood and double registration can't be caught by current `static register()` method.

Reply:
Though, after the change, client still cannot call `register` directly, because the singleton is private.

Reply:
Ok. I really dislike the contortions we have to go through in order to make this work as a static singleton, but thanks for explaining.

Reply:
The reason I changed the `static register()` to instance `register()` method is stated here. Also I should point out that because StormMetricsRegistry is a static singleton, it will persist throughout each UnitTest. So if a UnitTest class has multiple tests in it, all metrics will actually be registered multiple times. We either have to change all unit tests to have set up and tear down, or we have to completely revamp the StormMetricsRegistry class to make it non-static. As I don't have that much time to do either of them, I came up with this walkaround. @srdo @revans2 @Ethanlm

Reply:
I think we can make the registry non-static without it being a huge hassle https://github.com/apache/storm/pull/2783. I'll rebase it once https://github.com/apache/storm/pull/2752 goes in, and hopefully we can use it to resolve https://github.com/apache/storm/pull/2714.

Regarding the subclassing, I'm wondering if we should instead go ask the Dropwizard guys whether the inability to use `getOrAdd` for MetricSets is intended, and if not whether they'd be open to adding a method to MetricRegistry to allow it. If we're hitting this issue, other people probably are too, and I'd much rather fix it in the library than try to hack around it here.

What do you think @zd-project?

Reply:
I'll take a look at this

Reply:
Our customization upon MetricsRegistry is actually very similar to the wrapping methods of `#getOrAdd`, such as `gauge`, `timer`, `meter`, and `histogram`. They do not have a way to eliminate double registration of MetricSet though, although we could avoid this altogether with a non-static registry. 

In addition, I think we should probably upgrading to a newer version of Dropwizard, since current version (3.1.0) is about to be EOL. Their 4.x has a lot of improvement and provides more features on top of Java 8.

Reply:
Yes, I agree that for the non-MetricSet metrics, we can just use the getOrAdd wrappers. If we don't need MetricSet with a non-static registry, we should be good if we merge the changes in https://github.com/apache/storm/pull/2783. 

I agree that we should upgrade, but versions past 4.x have removed the metrics-ganglia module. I'm not sure if it's been spun off somewhere, or if it's just been deleted, but I didn't want to start removing stuff related to Ganglia in https://github.com/apache/storm/pull/2783 as well. If we want to upgrade past 3.1 I think we should do it in a separate PR.